crates/miden-objects/src/account/storage/map/witness.rs

```diff
@@ -1,50 +1,112 @@
+use alloc::collections::BTreeMap;
+
 use miden_crypto::merkle::{InnerNodeInfo, SmtProof};
 
 use crate::Word;
+use crate::account::StorageMap;
+use crate::errors::StorageMapError;
 
 /// A witness of an asset in a [`StorageMap`](super::StorageMap).
 ///
 /// It proves inclusion of a certain storage item in the map.
+///
+/// ## Guarantees
+///
+/// This type guarantees that the original key-value pairs it contains are all present in the
+/// contained SMT proof. Note that the inverse is not necessarily true. The proof may contain more
+/// entries than the map because to prove inclusion of a given original key A an
+/// [`SmtLeaf::Multiple`](miden_crypto::merkle::SmtLeaf::Multiple) may be present that contains both
+/// keys hash(A) and hash(B). However, B may not be present in the key-value pairs and this is a
+/// valid state.
 #[derive(Debug, Clone, PartialEq, Eq)]
-pub struct StorageMapWitness(SmtProof);
+pub struct StorageMapWitness {
+    proof: SmtProof,
+    map: BTreeMap<Word, Word>,
+}
 
 impl StorageMapWitness {
     // CONSTRUCTORS
     // --------------------------------------------------------------------------------------------
 
-    /// Creates a new [`StorageMapWitness`] from an SMT proof.
-    pub fn new(smt_proof: SmtProof) -> Self {
-        Self(smt_proof)
+    /// Creates a new [`StorageMapWitness`] from an SMT proof and a provided set of map keys.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if:
+    /// - Any of the map keys is not contained in the proof.
+    pub fn new(
+        proof: SmtProof,
+        map_keys: impl IntoIterator<Item = Word>,
+    ) -> Result<Self, StorageMapError> {
+        let mut map = BTreeMap::new();
+
+        for map_key in map_keys.into_iter() {
+            let hashed_map_key = StorageMap::hash_key(map_key);
+            let value =
+                proof.get(&hashed_map_key).ok_or(StorageMapError::MissingKey { map_key })?;
+            map.insert(map_key, value);
+        }
+
+        Ok(Self { proof, map })
+    }
+
+    /// Creates a new [`StorageMapWitness`] from an SMT proof and a set of key value pairs.
+    ///
+    /// # Warning
+    ///
+    /// This does not validate any of the guarantees of this type. See the type-level docs for more
+    /// details.
+    pub fn new_unchecked(
+        proof: SmtProof,
+        key_values: impl IntoIterator<Item = (Word, Word)>,
+    ) -> Self {
+        Self {
+            proof,
+            map: key_values.into_iter().collect(),
+        }
     }
 
     // PUBLIC ACCESSORS
     // --------------------------------------------------------------------------------------------
 
-    /// Searches for a value in the witness with the given `map_key`.
-    pub fn find(&self, hashed_map_key: Word) -> Option<Word> {
-        self.entries()
+    /// Returns the underlying [`SmtProof`].
+    pub fn as_proof(&self) -> &SmtProof {
```

Comment:
nit: since we are just returning a reference, I'd probably name this just `proof()`.